src/leader_finality_stage.rs

```diff
@@ -0,0 +1,198 @@
+//! The `leader_finality_stage` module implements the stage which regularly
+//! calculates the last finality times observed by the leader
+
+use bank::Bank;
+use influx_db_client as influxdb;
+use metrics;
+use service::Service;
+use std::result;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::thread::sleep;
+use std::thread::{self, Builder, JoinHandle};
+use std::time::Duration;
+use timing;
+use vote_program::VoteProgram;
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum FinalityError {
+    NoValidSupermajority,
+}
+
+pub const COMPTUTE_FINALITY_MS: u64 = 1000;
+
+pub struct LeaderFinalityStage {
+    compute_finality_thread: JoinHandle<()>,
+}
+
+impl LeaderFinalityStage {
+    fn get_last_supermajority_timestamp(
+        bank: &Arc<Bank>,
+        now: u64,
+        last_valid_validator_timestamp: u64,
+    ) -> result::Result<u64, FinalityError> {
+        let mut total_stake = 0;
+
+        let mut ticks_and_stakes: Vec<(u64, i64)> = {
+            let bank_accounts = bank.accounts.read().unwrap();
+            // TODO: Doesn't account for duplicates since a single validator could potentially register
```

Comment:
same: reference a github issue for TODOs, especially TODO comments that link to other TODO comments!